types/src/zkid/groth16_vk.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright © Aptos Foundation
+
+use crate::{move_utils::as_move_value::AsMoveValue, serialize};
+use aptos_crypto::CryptoMaterialError;
+use ark_bn254::{Bn254, G1Affine, G2Affine};
+use ark_groth16::{PreparedVerifyingKey, VerifyingKey};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use move_core_types::{
+    ident_str,
+    identifier::IdentStr,
+    move_resource::MoveStructType,
+    value::{MoveStruct, MoveValue},
+};
+use serde::{Deserialize, Serialize};
+use std::fmt::{Display, Formatter};
+
+/// Reflection of aptos_framework::zkid::Groth16PreparedVerificationKey
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Groth16VerificationKey {
+    pub alpha_g1: Vec<u8>,
+    pub beta_g2: Vec<u8>,
+    pub gamma_g2: Vec<u8>,
+    pub delta_g2: Vec<u8>,
+    pub gamma_abc_g1: Vec<Vec<u8>>,
+}
+
+impl AsMoveValue for Groth16VerificationKey {
+    fn as_move_value(&self) -> MoveValue {
+        MoveValue::Struct(MoveStruct::Runtime(vec![
+            self.alpha_g1.as_move_value(),
+            self.beta_g2.as_move_value(),
+            self.gamma_g2.as_move_value(),
+            self.delta_g2.as_move_value(),
+            self.gamma_abc_g1.as_move_value(),
+        ]))
+    }
+}
+
+/// WARNING: This struct uses resource groups on the Move side. Do NOT implement OnChainConfig
+/// for it, since `OnChainConfig::fetch_config` does not work with resource groups (yet).
+impl MoveStructType for Groth16VerificationKey {
+    const MODULE_NAME: &'static IdentStr = ident_str!("zkid");
+    const STRUCT_NAME: &'static IdentStr = ident_str!("Groth16VerificationKey");
+}
+
+impl TryFrom<Groth16VerificationKey> for PreparedVerifyingKey<Bn254> {
```

Comment:
nit: worth having some UTs for all the `impl TryFrom` across the PR.
If they are already covered by some smoke tests, then it's fine.

Reply:
Worth having a lot more tests! Next PR!